text/0000-partial_types.md

```diff
@@ -0,0 +1,672 @@
+
+I propose to extend type system by adding type integrity to sub-type. So, our variable will have next full type:
+```rust
+// case (A4)
+<variable>  : <type_clarification> <%integrity> <type>;
+<variable>  : <sharing> <'lifetime> <mutability> <%integrity> <type>;
+
+// case (A5)
+// FROM case (A1)
+foo  : value 'b const %a i16;
+foo  : value 'static const %full i16;
+foo  : %full i16;
+foo  : i16;
+```
+
+Lifetime variants are `'static` (for static lifetime), `'_`(don't care lifetime) and any other `'b`(some "b" lifetime).
+
+By the same analogy, integrity has similar names and meanings: `%full`(full integrity, soft keyword), `%_`(don't care how partial integrity is, soft keyword), `%empty` or `%!` (no integrity, soft keyword) and any other `%a`(some "a" integrity).
+
+Symbol `%` percent mean percent or part of the whole thing (variable in our case).
+
+_Note_: It is highly recommended to deprecate operator `%` as a remainder function (it is still no ambiguities to write "`\s+%\s+`"), and replace it with another operator (for example: `%mod` / `%rem` / `mod` / `rem`) to not to be confused by type integrity. 
+
+
+## Traits with Integrity variants
+
+We could already write Traits with **safe** virtual functions, that consumes virtual partial types having only variants of type integrity in Trait declaration
+```rust
+// case (B1)
+pub trait Upd {
+    type UpdType;
+
+    fn summarize<%a>(&self: & %a Self) -> String;
+
+    fn update_value<%a>(&mut self : &mut %a Self, newvalue: UpdType);
+
+    fn update_sametype<%a, %b>(&mut self : &mut %a Self, &another: & %b Self);
+}
+```
+
+Unfortunately, having variants of type integrity is not enough to write **safe** implementations or other non-virtual function declarations.
+
+
+## Detailed Integrity
+
+We need detailed integrity to write non-virtual specific typed parameters in function, including trait implementation.
+
+An abstractions is added for integrity detailing, we assume that **every** variable is a `struct`-like objects (even if it is not).
+
+We need for this some new quasi-fields and some field integrity (which should be soft keywords).
+
+_Note: I do not comment types (as is needed for compiler) for colorizing purposes._
+
+_Note: I use symbol `~` as a synonym of "equivalent" word only. It is not a Rust operator._
+```rust
+// case (C1)
+let foo = 0i16;
+    foo : i16  ~  %full i16;
+    foo : %full i16  ~  %{self.*} i16;
+    foo : %{self.*} i16  ~  %{self.self} i16;
+```
+
+Where :
+ - `self` is an "access" to variable itself
+ - `.*` is an "every field" quasi-field
+ - `.self` quasi-field for primitive types, since `i16` is not a `struct` type
+
+What's about structures? Almost the same:
+```rust
+struct Point {
+    x: f64,
+    y: f64,
+    z: f64,
+    t: f64,
+    w: f64,
+}
+
+// case (C2)
+let &mut p1 : &mut Point = Point {x:1.0, y:2.0, z:3.0, t:4.0, w:5.0};
+    p1 : &mut Point  ~  &mut %full Point;
+    p1 : &mut %{self.*} Point;
```

Comment:
Is there a reason why the outer braces are required if there's only one field? For example, I understand why you need braces for `%{self.x, self.y}`, but it makes sense to me that you can just write `%self.x` without the braces instead of `%{self.x}`.

Reply:
In vast majority of useful cases 2 fields are used `%{Self::field, %any}` or  `%{Self::field, %cut}`